File: solutions/169.Majority_Element/AC_Bits_n.py
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_Bits_n.py
# Create Date: 2015-02-05 11:30:25
# Descripton:   

class Solution:
    # @param num, a list of integers
    # @return an integer
    def majorityElement(self, num):
        res = 0
        for dig in range(32):
            app = 0
            for n in num:
                if (n & (1<<dig)):
                    app += 1
            if app * 2 > len(num):
                res |= (1<<dig)
            # Python will convert int to long
            # see https://oj.leetcode.com/discuss/21854/any-solution-using-python-to-count-bits
            if res & 0x80000000:
                res = -((res^0xffffffff) + 1)
        return res


# debug
s = Solution()
print s.majorityElement([-2147483648])

File: solutions/169.Majority_Element/AC_Bits_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_Bits_n.cpp
*  Create Date: 2015-02-05 10:49:09
*  Descripton:  Count 32 bits.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    int majorityElement(vector<int> &num) {
        int bitcnt[32] = {0};
        // swap the follow two 'for', it can work without bitcnt[]
        for (auto &i : num) {
            for (int j = 0; j < 32; ++j)
                if (i & (1<<j))
                    ++bitcnt[j];
        }
        int ans = 0;
        for (int i = 0; i < 32; ++i)
            if (bitcnt[i] > num.size() / 2)
                ans |= (1<<i);
        return ans;
    }
};

int main() {
    vector<int> num{-1, 1, 1};
    Solution s;
    cout << s.majorityElement(num) << endl;
    return 0;
}

